### PR TITLE
[refactor][doc] Improve schema docs

### PR DESCRIPTION
**Pull request by momo-jun**
### Motivation

This is a follow-up PR to https://github.com/apache/pulsar/pull/18242 to improve the Schema docs, aiming to provide Pulsar users with a more accurate, easy-to-read, and comprehensive information experience.

### Modifications

The content is revised and reviewed through [Google docs](https://docs.google.com/document/d/1YYansXopyVV66NaQ6TR9tfpq9MEhcbvCqv8ByFWuadA/edit?usp=sharing).
1. Improve the description and illustrations of `How Pulsar schema works`.
2. Add an illustration to introduce Pulsar schema.
3. Incorporate `schema evolution` and subtopics into `Understand schema`.
4. Relocate `native Avro` to `struct` schema.
5. Add incremental changes to align with the code.
6. Add/correct descriptions, commands, and code examples, e.g., fixing https://github.com/apache/pulsar/issues/15035
7. Remove duplicate or redundant information.

Many thanks to @congbobo184, @liangyepianzhou and @RobertIndie for providing technical inputs and review comments.

//cc @Anonymitaet @DaveDuggins @D-2-Ed 

### Documentation

- [x] `doc` 
- [ ] `doc-required` 
- [ ] `doc-not-needed` 
- [ ] `doc-complete`

### CODE REVIEW

File: site2/docs/schema-understand.md [hunk 1]
```diff
@@ -146,67 +159,88 @@ Producer<SensorReading> producer = client.newProducer(JSONSchema.of(SensorReadin
...
+|  `ALWAYS_INCOMPATIBLE`  |   N/A  | The schema evolution is disabled.                                                                                                                                                                                                                                                                                   | 
+|  <li>`BACKWARD` </li><li>`BACKWARD_TRANSITIVE` </li> |   Consumer first  | There is no guarantee that consumers using the old schema can read data produced using the new schema. Consequently, **upgrade all consumers first**, and then start producing new data.                                                                                                                            | 
+|  <li>`FORWARD` </li><li>`FORWARD_TRANSITIVE` </li> |   Producer first  | There is no guarantee that consumers using the new schema can read data produced using the old schema. Consequently, **upgrade all producers first**<li>to use the new schema and ensure that the data already produced using the old schemas are not available to consumers, and then upgrades the consumers. </li> |
```
Comment (labuladong) [inline site2/docs/schema-understand.md:245]:
```suggestion
|  <li>`FORWARD` </li><li>`FORWARD_TRANSITIVE` </li> |   Producer first  | There is no guarantee that consumers using the new schema can read data produced using the old schema. Consequently, **upgrade all producers first** to use the new schema and ensure that the data already produced using the old schemas are not available to consumers, and then upgrades the consumers. | 
```

there is a redundant list tag

File: site2/docs/schema-get-started.md [hunk 2]
```diff
@@ -11,80 +11,283 @@ import TabItem from '@theme/TabItem';
...
+<Tabs groupId="api-choice"
+  defaultValue="Java"
+  values={[{"label":"Java","value":"Java"},{"label":"C++","value":"C++"},{"label":"Python","value":"Python"},{"label":"Go","value":"Go"}]}>
+
+<TabItem value="Java">
+
+```java
+       .topic("my-topic")
+       .create();
+Consumer<byte[]> consumer = pulsarClient.newConsumer(Schema.BYTES)
```
Comment (labuladong) [inline site2/docs/schema-get-started.md:43]: Seems missing some code here

File: site2/docs/schema-get-started.md [hunk 3]
```diff
@@ -11,80 +11,283 @@ import TabItem from '@theme/TabItem';
...
+schema=BytesSchema())
+producer.send(b"Hello")
+
+consumer = client.subscribe(
+'bytes-schema-topic',
+		'sub',
+		schema=BytesSchema())
+msg = consumer.receive()
+data = msg.value()
+```
```
Comment (labuladong) [inline site2/docs/schema-get-started.md:83]: Indent style can be improved here.

File: site2/docs/schema-get-started.md [hunk 4]
```diff
@@ -11,80 +11,283 @@ import TabItem from '@theme/TabItem';
...
+  Value: []byte("message"),
+})
+
+consumer, err := client.Subscribe(pulsar.ConsumerOptions{
+  Topic:            "my-topic",
+  Schema:           pulsar.NewBytesSchema(nil),
+  SubscriptionName: "my-sub",
+  Type:             pulsar.Exclusive,
+})
+```
```
Comment (labuladong) [inline site2/docs/schema-get-started.md:103]: Suggest 4 space or a tab as indent for go code.

File: site2/docs/schema-get-started.md [hunk 5]
```diff
@@ -11,80 +11,283 @@ import TabItem from '@theme/TabItem';
...
+  Value: "message",
+})
+
+consumer, err := client.Subscribe(pulsar.ConsumerOptions{
+  Topic:            "my-topic",
+  Schema:           pulsar.NewStringSchema(nil),
+  SubscriptionName: "my-sub",
+  Type:             pulsar.Exclusive,
+})
+msg, err := consumer.Receive(context.Background())
```
Comment (labuladong) [inline site2/docs/schema-get-started.md:178]: Same suggestion as above.

File: site2/docs/schema-get-started.md [hunk 6]
```diff
@@ -164,12 +380,14 @@ You can define the `schemaDefinition` to generate a `struct` schema.
...
-   @AllArgsConstructor
-   @NoArgsConstructor
    public static class User {
-       String name;
-       int age;
+       public String name;
+       public int age;
+       public User(String name, int age) {
+ 	this.name = name;
+	this.age = age
```
Comment (labuladong) [inline site2/docs/schema-get-started.md:388]: Suggest all indents to be the same (4 spaces).

File: site2/docs/schema-get-started.md [hunk 7]
```diff
@@ -178,10 +396,10 @@ You can define the `schemaDefinition` to generate a `struct` schema.
    ```java
    SchemaDefinition<User> schemaDefinition = SchemaDefinition.<User>builder().withPojo(User.class).build();
    Producer<User> producer = client.newProducer(Schema.AVRO(schemaDefinition)).create();
-   producer.newMessage().value(User.builder().name("pulsar-user").age(1).build()).send();
+   producer.newMessage().value(new User (“pulsar-user”, 1)).send();
```
Comment (labuladong) [inline site2/docs/schema-get-started.md:399]:
```suggestion
   producer.newMessage().value(new User("pulsar-user", 1)).send();
```

Full-width quotes would be only used in Chinese.

File: site2/docs/schema-get-started.md [hunk 8]
```diff
@@ -228,88 +452,349 @@ Producer<SensorReading> producer = client.newProducer(AvroSchema.of(SensorReadin
...
+r = Example(a=1, b=2)
+producer.send(r)
 
-* Generate Protobuf schemas using `Schema.PROTOBUF`. The following example shows how to create the Protobuf schema and use it to instantiate a new producer:
+consumer = client.subscribe(
+'avro-schema-topic',
+		'sub',
+		schema=AvroSchema(Example))
+msg = consumer.receive()
+e = msg.value()
```
Comment (labuladong) [inline site2/docs/schema-get-started.md:503]: Same indent style problem as above.

File: site2/docs/schema-get-started.md [hunk 9]
```diff
@@ -228,88 +452,349 @@ Producer<SensorReading> producer = client.newProducer(AvroSchema.of(SensorReadin
...
 
+Suppose you have an `avroExampleStruct` class as follows, and you'd like to transmit it over a Pulsar topic.
 
-### Avro schema using C++
+```go
+  type avroExampleStruct struct {
+  ID   int
+  Name string
+}
+```
```
Comment (labuladong) [inline site2/docs/schema-get-started.md:581]: Same indent style problem as above.

**Review by labuladong — COMMENTED** [review]
Thanks for your work! Leave some style suggestions in the comments.

File: site2/docs/schema-get-started.md (hunk 2 above)
Comment (momo-jun) [reply site2/docs/schema-get-started.md]: @liangyepianzhou can you pls revisit this code snippet?

Comment (labuladong) [reply site2/docs/schema-get-started.md]:
```java
        Producer<byte[]> producer = client.newProducer(Schema.BYTES)
```
I think adding this part would be ok.

**Comment by momo-jun** [comment]
@labuladong thanks for your review. Comments have been addressed. PTAL.

@congbobo184 @liangyepianzhou @RobertIndie Do you have any other review comments?

**Review by liangyepianzhou — APPROVED** [review]

File: site2/docs/schema-understand.md [hunk 10]
```diff
@@ -48,92 +46,227 @@ Pulsar supports various schema types, which are mainly divided into two categori
...
+`ProtobufNativeSchema`|`ProtobufNativeSchema` is based on protobuf native Descriptor. <br /><br />This allows Pulsar to:<br />- use native protobuf-v3 to serialize or deserialize data<br />- use `AutoConsume` to deserialize data.|
+`Schema.NATIVE_AVRO` | `Schema.NATIVE_AVRO` is used to wrap a native Avro schema type `org.apache.avro.Schema`. The result is a schema instance that accepts a serialized Avro payload without validating it against the wrapped Avro schema. <br /><br />When you migrate or ingest event or messaging data from external systems (such as Kafka and Cassandra), the data is often already serialized in Avro format. The applications producing the data typically have validated the data against their schemas (including compatibility checks) and stored them in a database or a dedicated service (such as schema registry). The schema of each serialized data record is usually retrievable by some metadata attached to that record. In such cases, a Pulsar producer doesn't need to repeat the schema validation when sending the ingested events to a topic. All it needs to do is pass each message or event with its schema to Pulsar. |
```
Comment (RobertIndie) [inline site2/docs/schema-understand.md:100]: We can use `NativeAvroBytesSchema` instead of `Schema.NATIVE_AVRO` to make it consistent.

Comment (RobertIndie) [inline site2/docs/schema-understand.md:100]: I think it's better to link it to the Get started chapter for users to read more information.
`ProtobufNativeSchema` -> `schema-get-started/#protobufnative`
`Schema.NATIVE_AVRO` -> `schema-get-started/#native-avro`

**Review by RobertIndie — COMMENTED** [review]
Thanks for your contribution! I left some comments here.

File: site2/docs/schema-understand.md [hunk 11]
```diff
@@ -48,92 +46,227 @@ Pulsar supports various schema types, which are mainly divided into two categori
...
+| Primitive Type | Description | Java Type| Python Type | Go Type | C++ Type | C# Type|
+|---|---|---|---|---|---|---|
+| `BOOLEAN` | A binary value. | boolean | bool | bool | bool | bool |
+| `INT8` | A 8-bit signed integer. | int | int | int8 | int8_t | byte |
+| `INT16` | A 16-bit signed integer. | int | int | int16 | int16_t | short |
+| `INT32` | A 32-bit signed integer. | int | int | int32 | int32_t | int |
+| `INT64` | A 64-bit signed integer. | int | int | int64 | int64_t | long |
+| `FLOAT` | A single precision (32-bit) IEEE 754 floating-point number. | float | float | float32 | float | float |
+| `DOUBLE` | A double-precision (64-bit) IEEE 754 floating-point number. | double | double | float64| double | double |
+| `BYTES` | A sequence of 8-bit unsigned bytes. | byte[], ByteBuffer, ByteBuf | bytes | []byte | void * | byte[], `ReadOnlySequence<byte>` |
```
[the remark this answers is not in the document]
Comment (momo-jun) [reply site2/docs/schema-understand.md:58]: I used it as an escape from syntax errors. Without it, the compilation fails. The error message reads "Expected corresponding JSX closing tag for <byte>". Did you know any other ways to resolve it?

File: site2/docs/schema-understand.md [hunk 12]
```diff
@@ -48,92 +46,227 @@ Pulsar supports various schema types, which are mainly divided into two categori
...
-| `BYTES` | A sequence of 8-bit unsigned bytes | byte[], ByteBuffer, ByteBuf | bytes | []byte |
-| `STRING` | A Unicode character sequence | string | str | string| 
-| `TIMESTAMP` (`DATE`, `TIME`) |  A logic type represents a specific instant in time with millisecond precision. <br />It stores the number of milliseconds since `January 1, 1970, 00:00:00 GMT` as an `INT64` value |  java.sql.Timestamp (java.sql.Time, java.util.Date) | | |
-| INSTANT | A single instantaneous point on the time-line with nanoseconds precision| java.time.Instant | | |
-| LOCAL_DATE | An immutable date-time object that represents a date, often viewed as year-month-day| java.time.LocalDate | | |
-| LOCAL_TIME | An immutable date-time object that represents a time, often viewed as hour-minute-second. Time is represented to nanosecond precision.| java.time.LocalDateTime | |
-| LOCAL_DATE_TIME | An immutable date-time object that represents a date-time, often viewed as year-month-day-hour-minute-second | java.time.LocalTime | |
+| Primitive Type | Description | Java Type| Python Type | Go Type | C++ Type | C# Type|
+|---|---|---|---|---|---|---|
```
[the remark this answers is not in the document]
Comment (momo-jun) [reply site2/docs/schema-understand.md:50]: `N/A` is an abbreviation for `not applicable`. I think it's a common sense and can be easily googled :)

File: site2/docs/schema-understand.md (hunk 10 above)
Comment (momo-jun) [reply site2/docs/schema-understand.md]: Makes sense. I've added these links.

File: site2/docs/schema-understand.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (RobertIndie) [reply site2/docs/schema-understand.md:50]: Get it. Make sense.

File: site2/docs/schema-understand.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (RobertIndie) [reply site2/docs/schema-understand.md:58]: @momo-jun Hi, I have changed my code suggestions. And it works for me now.

**Review by congbobo184 — APPROVED** [review]

File: site2/docs/schema-get-started.md [hunk 13]
```diff
@@ -228,88 +329,350 @@ Producer<SensorReading> producer = client.newProducer(AvroSchema.of(SensorReadin
...
+       .topic("my-topic")
+       .subscriptionName("my-sub")
+       .subscribe();
+
+SchemaDema schemaDema = new SchemaDema();
+schemaDema.name = "puslar";
+schemaDema.age = 20;
+producer.newMessage().value(schemaDema).send();
+
+Message<SchemaDema> message = consumer.receive(5, TimeUnit.SECONDS);
```
[the remark this answers is not in the document]
Comment (momo-jun) [reply site2/docs/schema-get-started.md:529]: Good question. @liangyepianzhou can you pls take a look? I'm not sure whether it is intentional.

File: site2/docs/schema-understand.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (momo-jun) [reply site2/docs/schema-understand.md]: Great! It works! Thank you so much!

**Comment by momo-jun** [comment]
@RobertIndie @labuladong Thanks very much for your concrete review. Comments have been addressed. PTAL.

**Review by RobertIndie — APPROVED** [review]
LGTM. Thanks for your great work.

**Review by labuladong — APPROVED** [review]
Looks good to me~

**Pull request merged**